odl/tomo/data/uncompr_bin.py

```diff
@@ -0,0 +1,663 @@
+    allows to read arbitrary portions.
+    """
+
+    def __init__(self, file, header_fields=(), dtype=None, **kwargs):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        file : file-like or str
+            Stream or filename from which to read the data. The stream
+            is allowed to be already opened in ``'rb'`` mode.
+        header_fields : sequence of dicts, optional
+            Definition of the fields in the header (per row), each
+            containing key-value pairs for the following keys:
+
+            - ``'name'`` : Label for the field.
+            - ``'offset'`` : Start of the field in bytes.
+            - ``'size'`` : Size of the field in bytes.
+            - ``'dtype'`` : Data type in Numpy- or Numpy-readable format.
+            - ``'dshape'`` (optional) : The array of values is reshaped to
+              this shape.
+            - ``'description'`` (optional) : A human-readable description
+              of the field.
+
+            If empty, the header size is set 0, and the data type is
+            assumed to be ``dtype``. Use this in conjunction with
+            parametrized `read_data` to bypass the header and read
+            arbitrary data portions.
+
+        dtype : optional
+            Data type of the file's data block. It must be understood by
+            the `numpy.dtype` constructor. By default, the data type
+            is determined from the file header, or, if no information
+            is available there, it is set to ``np.dtype(float)``.
+        set_attrs : bool, optional
+            If ``True``, set attributes of ``self`` from the header for
+            convenient access. This can fail for non-standard
+            ``header_fields``, in which case ``False`` should be chosen.
+            Default: ``True``
+
+        See Also
+        --------
+        header_fields_from_table :
+            Function to parse a specification table, returning a field
+            sequence usable as ``header_fields`` parameter.
+        """
+        try:
+            file = open(file, 'rb')
+        except TypeError:
+            pass
+
+        if not file.readable() or 'b' not in file.mode:
+            raise ValueError("`file` must be opened readable in binary mode, "
+                             "but mode 'is {}'".format(file.mode))
+
+        self.__file = file
+        self.__set_attrs_from_header = bool(kwargs.pop('set_attrs', True))
+
+        # Initialize some attributes to default values, plus the header
+        self.data_shape = -1  # Makes reshape a no-op
+        self.data_dtype = np.dtype(dtype)
+        self.header_bytes = 0
+        self.__header_fields = header_fields
+        if self.header_fields:
+            self.__header = self.read_header()
+        else:
+            self.__header = None
+
+    @property
+    def file(self):
+        """File object from which ``self`` reads."""
+        return self.__file
+
+    def close(self):
```

Comment:
How about letting users use this class with `with`? Cleaner than requiring that they call this.